torch/csrc/jit/codegen/cuda/ir_internal_nodes.h

```diff
@@ -316,51 +355,96 @@ struct TORCH_CUDA_API TensorDomain : public Val {
   TensorDomain(TensorDomain&& other) = delete;
   TensorDomain& operator=(TensorDomain&& other) = delete;
 
-  TensorDomain(std::vector<IterDomain*> _domain)
-      : Val(ValType::TensorDomain), domain_(_domain) {}
+  TensorDomain(std::vector<IterDomain*> _domain);
+
+  TensorDomain(
+      std::vector<IterDomain*> _root_domain,
+      std::vector<IterDomain*> _domain);
+
+  TensorDomain(
+      std::vector<IterDomain*> _root_domain,
+      std::vector<IterDomain*> _rfactor_domain,
+      std::vector<IterDomain*> _domain);
 
   std::vector<IterDomain*>::size_type nDims() const {
     return domain_.size();
   }
 
   bool sameAs(const TensorDomain* const other) const;
 
+  static bool sameAs(
+      const std::vector<IterDomain*>& lhs,
+      const std::vector<IterDomain*>& rhs);
+
   const std::vector<IterDomain*>& domain() const noexcept {
     return domain_;
   }
 
   bool hasReduction() const;
-
+  bool hasBroadcast() const;
   bool hasRFactor() const;
 
-  TensorDomain* noReductions() const;
+  const std::vector<IterDomain*>& noReductions() const noexcept {
+    return no_reduction_domain_;
+  }
+
+  const std::vector<IterDomain*>& noBroadcasts() const noexcept {
+    return no_bcast_domain_;
+  }
+
+  const std::vector<IterDomain*>& rootDomain() const noexcept {
+    return root_domain_;
+  };
+
+  const std::vector<IterDomain*>& rfactorDomain() const noexcept {
+    return rfactor_domain_;
+  };
+
+  void resetDomains() {
+    no_reduction_domain_ = noReductions(domain_);
+    no_bcast_domain_ = noBroadcasts(domain_);
+  }
 
   // i here is int, as we want to accept negative value and ::size_type can be a
   // uint.
   IterDomain* axis(int i) const;
 
+  size_t posOf(IterDomain* id) const;
+
   // Split "axis" into 2 axes where the inner axes is size of "factor"
   // and outer axis is size axis.size() / factor
-  TensorDomain* split(int axis, int factor);
+  void split(int axis, unsigned int factor);
```

Comment:
why unsigned int? (unsigned types are more often than not a liability in interfaces and I think it would be a good idea to avoid them when possible)